boost: fix cross-compilation for Darwin

reckenrode opened the pull request:
Darwin needs a few tweaks to cross-compile Boost. This depends on https://github.com/NixOS/nixpkgs/pull/346043.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

reckenrode committed:
boost: fix Darwin cross-compilation

b2 tries to invoke the linker without a prefix, which fails.

reckenrode committed:
boost-build: fix aarch64 clang cross-compilation

SuperSandro2000 commented on pkgs/development/tools/boost-build/default.nix:
```diff
@@ -40,6 +40,9 @@ stdenv.mkDerivation {
   postPatch = ''
     substituteInPlace src/build-system.jam \
     --replace "default-toolset = darwin" "default-toolset = clang-darwin"
```
```suggestion
      --replace-fail "default-toolset = darwin" "default-toolset = clang-darwin"
```

reckenrode replied:
Latest push drops this replacement because it fails with `--replace-fail`. It appears upstream has made clang the default.

reckenrode commented:
The nixfmt-check errors are for files not included in this PR.

reckenrode committed:
boost-build: drop Darwin compiler substitution

Upstream defaults to clang on Darwin now instead of GCC. This
substitution doesn’t do anything.

emilazy commented on pkgs/development/tools/boost-build/default.nix:
```diff
@@ -36,10 +36,10 @@ stdenv.mkDerivation {
   postPatch = ''
-    substituteInPlace src/build-system.jam \
-    --replace "default-toolset = darwin" "default-toolset = clang-darwin"
+    # Upstream uses arm64, but nixpkgs uses aarch64.
+    substituteInPlace src/tools/clang.jam \
+      --replace-fail 'arch = arm64' 'arch = aarch64'
```
I wonder if we should consider adding aliases for this to our LLVM build at some point…

paparodeo replied:
seems like the replace-fail should be replace-warn as older boosts are breaking here. and think that the default-toolset should be added back due to older boosts as well, using `replace-warn`.

paparodeo replied:
see https://github.com/NixOS/nixpkgs/pull/350658

trofi commented:
Bisect says 90df0c3ed8794b60e95bb105674dd3f6d7f81972 `boost-build: fix aarch64 clang cross-compilation` broke `boost179` (and a few other boost versions) in` staging-next` and `staging` as:

```
$ nix build --no-link -f. boost179 -L
...
boost-build-boost> Running phase: unpackPhase
boost-build-boost> unpacking source archive /nix/store/p009j36gh69b1g748lg5gr0409w1a6z8-boost_1_79_0.tar.bz2
boost-build-boost> source root is boost_1_79_0
boost-build-boost> setting SOURCE_DATE_EPOCH to timestamp 1649279392 of file boost_1_79_0/tools/build/index.html
boost-build-boost> unpackPhase completed in 1 minutes 11 seconds
boost-build-boost> Running phase: patchPhase
boost-build-boost> substituteStream() in derivation boost-build-boost-1.79.0: ERROR: pattern arch\ =\ arm64 doesn't match anything in file 'src/tools/clang.jam'
```

reckenrode commented:
It also needs investigating to see whether any of those older versions of Boost still need the `default-toolset` replacement to use clang on Darwin.
